Implement light-weight Workspace Trust API for plugins

martin-fleck-at opened the pull request:
#### What it does
- Add trust-relevant methods and interfaces to the plugin workspace API
- Add request for workspace trust to workspace service
- Allow application setting to auto-respond to trust request

https://github.com/eclipse-theia/theia/issues/10472

**Note: I share this PR to showcase how the issue might be solved as I tried to double-check the issue I was facing. While it is working and hopefully not doing any harm, further discussions and improvements might be needed for a real support of the workspace trust API.**

#### How to test
- Create a Theia Electron application as described in https://github.com/eclipse-theia/theia/issues/10472
- Trigger the remote pickup through: Source Control View > Remote > Add Remote...

![fix_git](https://user-images.githubusercontent.com/19170971/143605667-f93685f6-3ada-4bd5-85af-d7e3da306298.png)
(working Remote query)

![git-trust](https://user-images.githubusercontent.com/19170971/143605941-7999ee23-f4f3-4400-ade7-554a3d3ef6b3.png)
(user trust query)

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

vince-fugnitto commented on packages/plugin-ext/src/main/browser/workspace-main.ts:
```diff
@@ -14,23 +14,23 @@
 import { URI as Uri } from '@theia/core/shared/vscode-uri';
-import { UriComponents } from '../../common/uri-components';
 import { FileSearchService } from '@theia/file-search/lib/common/file-search-service';
-import URI from '@theia/core/lib/common/uri';
-import { WorkspaceService } from '@theia/workspace/lib/browser';
-import { Resource } from '@theia/core/lib/common/resource';
-import { Disposable, DisposableCollection } from '@theia/core/lib/common/disposable';
-import { Emitter, Event, ResourceResolver, CancellationToken } from '@theia/core';
-import { PluginServer } from '../../common/plugin-protocol';
 import { FileSystemPreferences } from '@theia/filesystem/lib/browser';
-import { SearchInWorkspaceService } from '@theia/search-in-workspace/lib/browser/search-in-workspace-service';
 import { FileStat } from '@theia/filesystem/lib/common/files';
 import { MonacoQuickInputService } from '@theia/monaco/lib/browser/monaco-quick-input-service';
+import * as theia from '@theia/plugin';
+import { SearchInWorkspaceService } from '@theia/search-in-workspace/lib/browser/search-in-workspace-service';
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import { MAIN_RPC_CONTEXT, StorageExt, WorkspaceExt, WorkspaceFolderPickOptionsMain, WorkspaceMain } from '../../common/plugin-api-rpc';
+import { PluginServer } from '../../common/plugin-protocol';
+import { RPCProtocol } from '../../common/rpc-protocol';
+import { UriComponents } from '../../common/uri-components';
```
@martin-fleck-at do you mind reverting these unnecessary changes, it makes the pull-request harder to review.

martin-fleck-at replied:
Ah yes, sorry, I will push an update for that soon.

vince-fugnitto commented on packages/plugin-ext/src/plugin/workspace.ts:
```diff
@@ -56,6 +56,10 @@ export class WorkspaceExtImpl implements WorkspaceExt {
     private searchInWorkspaceEmitter: Emitter<{ result?: theia.TextSearchResult, searchId: number }> = new Emitter<{ result?: theia.TextSearchResult, searchId: number }>();
     protected workspaceSearchSequence: number = 0;
 
+    private _trusted: boolean = false;
```
I believe we should persist the choice from the user, unfortunately on each startup the application will ask users which can become annoying.

martin-fleck-at replied:
Great idea, I store the value now if the preference is set to ask only 'once'.

colin-grant-work commented on packages/workspace/src/browser/workspace-service.ts:
```diff
@@ -111,6 +113,11 @@ export class WorkspaceService implements FrontendApplicationContribution {
             }
         });
+
+        const autoTrust = FrontendApplicationConfigProvider.get()['autoTrust'];
+        if (typeof autoTrust === 'boolean') {
+            this.trustRequest.resolve(autoTrust);
+        }
```
I know the headline says 'light-weight,' but I'd prefer to also implement some of the [preferences](https://code.visualstudio.com/docs/editor/workspace-trust#_settings) associated with the feature so that users have some control.

martin-fleck-at replied:
Good idea! I have implemented some of the settings in a follow-up change and rebased my branch.

colin-grant-work commented on packages/workspace/src/browser/workspace-trust-service.ts:
```diff
@@ -0,0 +1,115 @@
+        }
+    }
+
+    get isTrusted(): boolean | undefined {
+        return this._isTrusted;
+    }
```
I don't see any references to this field elsewhere?

martin-fleck-at replied:
Right, it was a leftover of some experiment, sorry for that!

colin-grant-work commented on packages/workspace/src/browser/workspace-trust-service.ts:
```diff
@@ -0,0 +1,115 @@
+        return this._isTrusted;
+    }
+
+    get onTrusted(): Promise<boolean> {
+        return this.trusted.promise;
+    }
```
The `on__` name convention is usually reserved for events. The way this is used has some similarity to an event (fire once, check often), but I'd prefer either just changing this to `get isTrusted` (since that isn't currently used) or something like
```typescript
getWorkspaceTrust(): Promise<Boolean> {
    return this.trusted.promise;
}
}

martin-fleck-at replied:
You are right, I made a mistake there. Fixed it and went with the getWorkspaceTrust approach.

colin-grant-work commented on packages/workspace/src/browser/workspace-trust-service.ts:
```diff
@@ -0,0 +1,115 @@
+    protected async init(): Promise<void> {
+        await this.workspaceService.ready;
+        this.trusted.promise.then(trusted => this._isTrusted = trusted);
+        const implicitTrust = await this.calculateWorkspaceTrust();
+        if (implicitTrust !== undefined) {
+            this.trusted.resolve(implicitTrust);
+        }
```
Does this need to be run when this class is initialized, or could we wait until the first call to `requestWorkspaceTrust`?

martin-fleck-at replied:
Ideally yes, since in the plugin API the 'onDidGrantWorkspaceTrust' which we trigger after resolving this sets the trusted flag on the workspace (see plugin-context.ts and workspace.ts in plugin-ext). The Git plugin, for instances, has code like this:
```
if (workspace.isTrusted) {
   this.disposables.push(new AutoFetcher(this, globalState));
} else {
   const trustDisposable = workspace.onDidGrantWorkspaceTrust(() => {
      trustDisposable.dispose();
      this.disposables.push(new AutoFetcher(this, globalState));
   });
   this.disposables.push(trustDisposable);
}
```
So they query the flag early on and afterwards listen to any potential changes. Therefore we should set and propagate that flag as early as possible if we already know what the value will be. What do you think?

colin-grant-work replied:
Sounds fair to me - I'm not very familiar with the plugin side of things, but if they're expecting that it might be resolved before they're activated, then we should make the effort to satisfy that expectation. 👍

colin-grant-work commented on packages/workspace/src/browser/workspace-trust-service.ts:
```diff
@@ -0,0 +1,115 @@
+        return undefined;
+    }
+
+    async requestWorkspaceTrust(): Promise<boolean | undefined> {
+        if (this.trusted.state === 'unresolved') {
+            const queryTrusted = await this.messageService.info(nls.localize('theia/workspace/trustRequest',
+                'An extension requests workspace trust but the corresponding API is not yet fully supported. Do you want to trust this workspace?'),
+                Dialog.YES, Dialog.NO);
+            const trusted = queryTrusted === Dialog.YES;
+
+            const trustPrompt = this.workspaceTrustPref[WORKSPACE_TRUST_STARTUP_PROMPT];
+            if (trustPrompt === WORKSPACE_TRUST_STARTUP_PROMPT_ONCE) {
+                await this.storage.setData(STORAGE_TRUSTED, trusted);
+            }
+            this.trusted.resolve(trusted);
+        }
+        return this.trusted.promise;
+    }
```
If it's reasonable to wait to run `calculateWorkspaceTrust` until the first call to `requestWorkspaceTrust`, it would be better to have the query as the final fallback if we arrive at the end of `calculateWorkspaceTrust` without having returned a `boolean` value.

martin-fleck-at replied:
I agree but I think we need to calculate the flag early on if possible, see my previous answer.

martin-fleck-at commented on packages/workspace/src/browser/workspace-trust-service.ts:
```diff
@@ -0,0 +1,115 @@
+/********************************************************************************
+ * Copyright (C) 2017 TypeFox and others.
```
Thank you, done!

martin-fleck-at commented on packages/workspace/src/browser/workspace-trust-preferences.ts:
```diff
@@ -0,0 +1,77 @@
+/********************************************************************************
+ * Copyright (C) 2018 Ericsson and others.
```
Thank you, done!

martin-fleck-at commented on packages/workspace/src/browser/workspace-trust-service.ts:
```diff
@@ -0,0 +1,115 @@
+
+    protected async calculateWorkspaceTrust(): Promise<boolean | undefined> {
+        await this.preferences.ready;
+        const autoTrust = FrontendApplicationConfigProvider.get()[CONFIG_AUTO_TRUST];
```
I was not aware of this approach, I adapted the code accordingly. Thank you for pointing that out!

colin-grant-work commented on packages/workspace/src/browser/workspace-trust-service.ts:
```diff
@@ -0,0 +1,101 @@
+
+        const trustPrompt = this.workspaceTrustPref[WORKSPACE_TRUST_STARTUP_PROMPT];
+        if (trustPrompt === WORKSPACE_TRUST_STARTUP_PROMPT_ONCE) {
+            return this.storage.getData<boolean>(STORAGE_TRUSTED);
+        }
+
+        // reset stored trusted setting
+        this.storage.setData(STORAGE_TRUSTED, undefined);
```
The handling of the storage is a bit odd. Basically, we only care about storing the value if the user has set `PROMPT_ONCE` - and we want to be able to save the value if the user changes that preference value after we run `requestWorkspaceTrust`, so perhaps the saving / deleting should be handled in a preference change listener, as well?

martin-fleck-at replied:
I will implement that in a follow-up change as well, I think that might be a good idea, thank you!

martin-fleck-at commented on packages/workspace/src/browser/workspace-trust-service.ts:
```diff
@@ -0,0 +1,101 @@
+        return this.trusted.promise;
+    }
+
+    protected async calculateWorkspaceTrust(): Promise<boolean | undefined> {
```
Yes, I would like to keep that for customization purposes if that is alright. If not I can change it back.

martin-fleck-at commented on packages/plugin-ext/src/plugin/workspace.ts:
```diff
@@ -418,4 +422,23 @@ export class WorkspaceExtImpl implements WorkspaceExt {
         }
     }
+
+    get trusted(): boolean {
+        if (!this._trusted) {
+            this.requestWorkspaceTrust();
+        }
+        return this._trusted;
+    }
+
+    requestWorkspaceTrust(options?: theia.WorkspaceTrustRequestOptions): Promise<boolean | undefined> {
+        return this.proxy.$requestWorkspaceTrust(options);
+    }
```
You are right, that this is a bit odd. The main 'problem' is that we do not have a Promise here to return, so we cannot actually await the result of the request. In fact, now that I think about it, I don't think we should to do the request here at all. I double checked and [VS Code](https://github.com/microsoft/vscode/blob/a0a448f344bc8c068deb1202708dfc1e7c17ef89/src/vs/workbench/api/common/extHostWorkspace.ts#L561) is also not doing it implicitly, so I removed this call for now. That means that plugins need to explicitly call the 'requestWorkspaceTrust' method if they require it.

martin-fleck-at replied:
We now only return the trusted value and skip the async request as discussed above.

martin-fleck-at commented on packages/workspace/src/browser/workspace-trust-service.ts:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    protected async calculateWorkspaceTrust(): Promise<boolean | undefined> {
+        if (!this.workspaceTrustPref[WORKSPACE_TRUST_ENABLED]) {
```
Yeah, I think you are right about the setting. I will add a change that listens to preference changes and updates the trust only if the value has not already been set to avoid such unexpected behavior. However, I don't think we should do that if the trust has already been set since that might be even more confusing, what do you think?

colin-grant-work commented on packages/workspace/src/browser/workspace-trust-preferences.ts:
```diff
@@ -0,0 +1,84 @@
+        [WORKSPACE_TRUST_STARTUP_PROMPT]: {
+            description: nls.localizeByDefault('Controls when the startup prompt to trust a workspace is shown.'),
+            enum: Object.values(WorkspaceTrustPrompt),
+            default: typeof FrontendApplicationConfigProvider.get()[CONFIG_AUTO_TRUST] === 'boolean' ? WorkspaceTrustPrompt.NEVER : WorkspaceTrustPrompt.ALWAYS
```
Here the norm would be to implement a default in the `preferences` section of the application configuration rather than introducing a separate field. If you'd like to implement a separate field, you probably want to check the value of the field, not just its type.

martin-fleck-at replied:
You are right, we can simply override the preferences in the application configuration so there is no real need for a dedicated field. I changed it to "normal" default values.

colin-grant-work commented on packages/workspace/src/browser/workspace-trust-preferences.ts:
```diff
@@ -0,0 +1,84 @@
+        [WORKSPACE_TRUST_ENABLED]: {
+            description: nls.localize('theia/workspace/trustEnabled', 'Controls whether or not workspace trust is enabled. If disabled, all workspaces are trusted.'),
+            type: 'boolean',
+            default: FrontendApplicationConfigProvider.get()[CONFIG_AUTO_TRUST] === false
```
Here, too.

martin-fleck-at replied:
Done

colin-grant-work commented on packages/workspace/src/browser/workspace-trust-preferences.ts:
```diff
@@ -0,0 +1,84 @@
+        [WORKSPACE_TRUST_EMPTY_WINDOW]: {
+            description: nls.localize('theia/workspace/trustEmptyWindow', 'Controls whether or not the empty workspace is trusted by default.'),
+            type: 'boolean',
+            default: FrontendApplicationConfigProvider.get()[CONFIG_AUTO_TRUST] === true
```
Here, too, use the default from VSCode and let application developers set their own defaults in the `preferences` section if desired.

colin-grant-work commented on packages/workspace/src/browser/workspace-trust-service.ts:
```diff
@@ -0,0 +1,119 @@
+    protected async init(): Promise<void> {
+        await this.workspaceService.ready;
+        await this.resolveWorkspaceTrust();
+        this.preferences.onPreferenceChanged(change => this.preferenceChanged(change));
```
```suggestion
        this.preferences.onPreferenceChanged(change => this.handlePreferenceChange(change));
```

martin-fleck-at replied:
Method was renamed

colin-grant-work commented on packages/workspace/src/browser/workspace-trust-service.ts:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    protected async resolveWorkspaceTrust(givenTrust?: boolean): Promise<void> {
+        if (this.workspaceTrust.state === 'unresolved') {
```
If the user switches the preference value from trust enabled `false` -> trust enabled `true`, what should happen? With the current code, we would have resolved the trust to `true` at startup, but now the user might be requesting that we rethink that. It seems that VSCode requires a restart if you change whether workspace trust is enabled or not, presumably so that it can rerun the checks that it would run at startup.

martin-fleck-at replied:
Great suggestion! You are right that this is a bit confusing. I now ask the user if they want to restart the application if they change the trust enablement, similar to what VS Code is doing. I think with this the implication of that preference should be much clearer.

martin-fleck-at commented:
@colin-grant-work I hope you had a great start into the new year and thank you very much for your extensive review! It took me some time to come back to this but I updated the PR as you suggested and everything should now be working as expected.

colin-grant-work commented on packages/workspace/src/browser/workspace-trust-preferences.ts:
```diff
@@ -0,0 +1,76 @@
+        [WORKSPACE_TRUST_ENABLED]: {
+            description: nls.localize('theia/workspace/trustEnabled', 'Controls whether or not workspace trust is enabled. If disabled, all workspaces are trusted.'),
+            type: 'boolean',
+            defaultValue: false
```
Is it your intent to make this disabled by default (i.e. make all workspaces trusted?)

martin-fleck-at replied:
Yes, that was the intent. I believe VS Code does it similarly, see https://github.com/microsoft/vscode/blob/f7bfe5ca6682f27bc0ef19b412a4888078265ecc/src/vs/workbench/services/workspaces/common/workspaceTrust.ts#L291

colin-grant-work reviewed (commented):
Something seems to be malfunctioning. I modified my frontend config to enable workspace trust (see comment below), confirmed that, on startup, workspace trust was not resolved, and then ran the `Add remote...` command.

When I selected the `Add remote from GitHub` option, I got this notice:

![image](https://user-images.githubusercontent.com/62660806/151029197-7ffec473-0e12-4504-8429-4b30ca1c4fc1.png)

That looks like a problem with the authentication API, not the workspace trust API (probably).

When I just put in the URL of the Theia repo, I was prompted to name the remote, and the process concluded. I was never prompted to confirm workspace trust, but I think I should have been? Or should the request only have come when using GitHub as an authentication provider?

martin-fleck-at commented:
@colin-grant-work That behavior is expected now that we no longer explicitly call `requestWorkspaceTrust` whenever a plugin queries the `isTrusted` flag. I had that originally but I think it was removed as part of the review https://github.com/eclipse-theia/theia/pull/10473#discussion_r761244478. So we either need to find a good trigger point to request the workspace trust, e.g., when an API registers a listener for it with `onDidGrantWorkspaceTrust` in which case the asynchronous nature does not hurt us or we expect the plugins to call that behavior explicitly which may be a bad idea seeing that the vscode git-plugin already does not do that. I suggest to add it on the `onDidGrantWorkspaceTrust` if the trusted flag is too risky, what do you think?

colin-grant-work commented:
@martin-fleck-at, that all sounds reasonable, and it explains why, the first time I ran through the test steps with workspace trust automatically enabled, I got an additional prompt asking me if I wanted to do automatic fetches, which is the feature that's checking workspace trust.

So as you say, the question is when to prompt, I guess. I find VSCode's prompt-on-start pretty annoying, so maybe your original idea of requesting trust when anyone checks for it is the best idea. The GitHub plugin immediately creates a listener if it finds the workspace is untrusted, and that seems like the only good way for a plugin to handle finding `isTrusted` === `false`. That has the benefit of not requesting trust if no one depends on it and requesting trust in a timely manner if a feature activates that does depend on it. In that case, though, we probably want to have a system that caches a definite negative response so that we don't pester the user with multiple requests if they've already denied it.

colin-grant-work commented:
I've implemented a tiny VSCode plugin [here](https://github.com/colin-grant-work/vscode-extension-samples/tree/demo/workspace-trust/workspace-api-test) with commands to exercise the workspace API.

martin-fleck-at commented:
@colin-grant-work Thank you very much for the plugin, it was very useful to test the different scenarios! As you suggested, we now request the workspace trust whenever the trust is queried. In addition to that, we report any workspace trust state back to the plugin (`onWorkspaceTrustChanged`) even when we determine that it is false. In general, the user should not be pestered with multiple requests anyway as once we resolve the workspace trust no further dialogs are shown. This is also the reason that we trigger a restart when the respective preference was changed, similar to VS Code. 

Thank you for your review and effort!

colin-grant-work reviewed (commented):
I'm happy with the code and the current behavior. The only outstanding question is whether to enable or disable the workspace trust feature by default, and I'd like to give others a chance to weigh in.

JonasHelming commented:
I think we should enable the feature by default as it provides a more defensive strategy when it comes to security. Adopters can still choose to disable the feature in their application default preferences if they accept any risk that may come with it. However, if they do not do that it is probably better to inform the user that a plugin requests workspace trust.

martin-fleck-at commented:
@JonasHelming @colin-grant-work Thank you for your feedback! Thinking about it a bit more, I fully agree with Jonas that if no plugin requests the trust, the user is not bothered anyway but if a plugin needs it, it may be better to inform the user unless some adopters explicitly choose not to do so. As such, I changed the default enablement preference to `true` and squashed and rebased the change. If there are any more changes necessary before this can be merged, please let me know.

JonasHelming commented:
@colin-grant-work : I would merge this now, OK?

msujew reviewed (commented):
@colin-grant-work I don't really have a strong opinion on that. I think both would be fine, but I'm slightly leaning towards not enabling it by default.

In the end, downstream users of Theia can always choose to have it enabled by default in their product.

JonasHelming commented:
Just to mention, I also do not have a strong opinion about it. So @colin-grant-work Your vote will decide on this, no option is "wrong"

colin-grant-work commented:
@msujew, @JonasHelming, @martin-fleck-at, thanks for your input. As annoying as I find VSCode's modal at start up, I think enabling security by default is better than disabling it by default - adopters and users are free to follow whatever policies their risk managers will let them get away with :-). So default on sounds good to me, and I'll merge.

colin-grant-work reviewed (approved):
The latest code looks good to me, and I think we've settled on enabling the feature by default.
